common/config.gni

```diff
@@ -58,6 +60,9 @@ if (is_ios || is_mac) {
     "-Werror=overriding-method-mismatch",
     "-Werror=undeclared-selector",
   ]
+  if (extension_safe) {
+    flutter_cflags_objc += [ "-fapplication-extension" ]
```

Comment:
macOS doesn't fail when building with this flag, but it doest contain [NSApplication code](https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm#L177). I'm confused here. 

@jmagman @cbracken

Reply:
I didn't get this working either, might take more investigation to see how what flags Xcode is passing through when it shows the warning.

Reply:
I tested the locally built mac engine.
Without with this flag, I got the warning when linking a share extension to the framework.
With this flag, there is no warning. 
So it seemed like it is just working (which clearly is not since the mac os code does contain NSApplication.)

Will investigate more. 
(I should have marked this PR as draft, it is not intended for reviewing yet before this macOS mystery is resolved)

Reply:
I looked into the headers, on iOS:

`@property(class, nonatomic, readonly) UIApplication *sharedApplication NS_EXTENSION_UNAVAILABLE_IOS("Use view controller based solutions where appropriate instead.");`
The shared application is marked as extension unavailable.

However, on macOS it is not.
`@property (class, readonly, strong) __kindof NSApplication *sharedApplication;` 

In the document: https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/ExtensionOverview.html#//apple_ref/doc/uid/TP40014214-CH2-SW6 only linked to the UIApplication.sharedApplication, so it sounds like NSApplication is ok?

I create a sample mac framework having `Require Only App-Extension-safe API` turned on and having NSApplication.SharedApplication. It successfully compiled and a share extension can link to it without warning.

So I believe NSApplication.sharedApplication is compatible with extensions?

Reply:
> So I believe NSApplication.sharedApplication is compatible with extensions? So on mac, as long as the `-fapplication-extension` is in the build flags and there is no compile error, it will work.

!!!
From my reading of the docs when they said `sharedApplication` without specifying an object I thought they meant either `UIApplication` or `NSApplication`.  

> Access a [sharedApplication](https://developer.apple.com/documentation/uikit/uiapplication/1622975-shared) object, and so cannot use any of the methods on that object

If we don't need to do anything other than documentation for macOS that would be amazing!

Reply:
And also you could pull that very simple macOS flag-only work out of this so it would start working in the next stable.

Reply:
By "next stable" I just mean it can be available now-ish without needing to wait for all the iOS work to be correct.